### DISCUSSION TITLE
AutoNAT: "refusing to dial peer with blocked observed address"

### DISCUSSION THREAD

**Discussion by Wiezzel** [libp2p/rust-libp2p#4529] [Q&A] [answered]
I am starting a node, connecting to another one, and reading logs from `libp2p_autonat`. There I see this message appearing and I would like to understand what it means. Why is my address 'blocked'?
```
2023-09-20T15:49:23Z DEBUG libp2p_autonat::behaviour::as_client] Outbound dial-back request returned DialResponse { status_text: Some("refusing to dial peer with blocked observed address"), result: Err(DialRefused) }.
```

**Comment by thomaseizinger** [comment, marked as answer]
It is either because you are connecting from a relayed address or one that is not in the global IP range like 192.168.x.x.

See https://github.com/libp2p/rust-libp2p/blob/854a6ae868bc782acdea6d0b49e78902b3f46223/protocols/autonat/src/behaviour.rs#L200.

**Reply by Wiezzel** [reply]
@thomaseizinger OK, so I have a follow-up question. The node that is sending the dial-back request is listening on `/ip4/95.216.112.42/tcp/12347`. That's the only listen addr that it uses. How is it possible that the remote peer recognizes it as connecting from a relayed or non-global address?

This is the `identify` protocol event logged on the remote peer.
```
Received { peer_id: PeerId("12D3KooWSyqmbHd1SYNnCNkJzanF37YnYD7zwG8WbqUSpuM15MQ5"), info: Info { public_key: PublicKey { publickey: Ed25519(PublicKey(compressed): ff41d62c6ff82ca262aeee78fca548caffe2436a442877425795a8abceceaa) }, protocol_version: "/subsquid/0.0.1", agent_version: "rust-libp2p/0.43.0", listen_addrs: ["/ip4/95.216.112.42/tcp/12347"], protocols: [StreamProtocol { inner: Right("/subsquid-worker/0.0.1") }, StreamProtocol { inner: Right("/ipfs/id/push/1.0.0") }, StreamProtocol { inner: Right("/meshsub/1.0.0") }, StreamProtocol { inner: Right("/meshsub/1.1.0") }, StreamProtocol { inner: Right("/libp2p/autonat/1.0.0") }, StreamProtocol { inner: Right("/ipfs/id/1.0.0") }, StreamProtocol { inner: Right("/libp2p/circuit/relay/0.2.0/stop") }], observed_addr: "/dns4/testnet-stage.subsquid.io/tcp/42346/p2p/12D3KooWQER7HEpwsvqSzqzaiV36d3Bn6DZrnwEunnzS76pgZkMU" } }
```
As you see, `listen_addrs` contains only a single public address.

**Reply by thomaseizinger** [reply]
It doesn't matter what the node is listening on but which address the AutoNAT server sees as the origin for the connection, i.e. it reads the IP address from the local TCP socket for example.

If the connection was established through a relay, then we track that internally and report that accordingly.

**Reply by Wiezzel** [reply]
@thomaseizinger 
1) Is it possible to see the observed address in the logs? Would `libp2p_autonat=debug` do it?
2) Is there a way to to force AutoNAT to probe the listen addr instead of the incoming connection addr?

**Reply by thomaseizinger** [reply]
> Is there a way to to force AutoNAT to probe the listen addr instead of the incoming connection addr?

That would turn all nodes into a bot-net that you can use to make DoS attacks :D

There is a V2 specification in the works that allows the remote to specify testing of an arbitrary testing in exchange for "paying" with some bandwidth: https://github.com/libp2p/specs/pull/538

**Reply by thomaseizinger** [reply]
> Is it possible to see the observed address in the logs? Would `libp2p_autonat=debug` do it?

I don't recall all exact logs but I am pretty sure some part of the stack will log inbound and outbound connections. You can always implement your own `NetworkBehaviour` and add logging for the specific callbacks: https://github.com/libp2p/rust-libp2p/blob/484fc5693c0a7ad9ad2acb0a08445a7a4f5f7fd9/swarm/src/behaviour.rs#L133-L191

**Comment by Wiezzel** [comment]
@thomaseizinger Thank you, sir! That was very helpful. 🙏🏻 ❤️
